### ISSUE TITLE
Variations as to how MSP saves custom field lookup values

### ISSUE THREAD

**Issue by ndarlington** [joniles/mpxj#114]
## MPXJ versions:
7.6.x - 7.9.2 inclusive

## MSP versions:
Seems to be at least any 2016-2019 standard or pro
- (specifically I'm using Project for Office 365, 16.0.11328.20362 - 32-bit right now but we have customers and testers on other versions and haven't found one that isn't affected yet).

## Problem summary:
When importing an .xml into MSP and saving it as .mpp, if there were lookup values on columns (e.g. Text1) for rows that had not been visibly shown on a view prior to saving it and reading it with MPXJ, it seems it stores the lookup data in a different place within the .mpp file and causes MPXJ to read garbage characters from something else instead.

## Test code for demonstration:
```java
package com.regoconsulting.mpxj;

import net.sf.mpxj.MPXJException;
import net.sf.mpxj.ProjectFile;
import net.sf.mpxj.Task;
import net.sf.mpxj.reader.UniversalProjectReader;

import java.io.IOException;

public class Main {

    public static void main(String[] args) throws MPXJException, IOException {
        if (args.length > 0)
            ReadTaskLookupValues(args[0].startsWith("com.rego") ? args[1] : args[0]);
    }

    private static void ReadTaskLookupValues(String filename) throws MPXJException, IOException {
        UniversalProjectReader reader = new UniversalProjectReader();
        ProjectFile project = reader.read(filename);

        for (Task task : project.getTasks()) {
            System.out.println(task.getName() + ": " + task.getText(1));
        }
    }
    
}
```

## Setup and steps to reproduce:
1. Create new project in MSP, add 1 task
2. Go to Custom Fields, select Lookup for Text1 (task) and enter a few dummy values like so:
![image](https://user-images.githubusercontent.com/40403963/63405471-8da98880-c3ac-11e9-93b0-95256b809c15.png)
3. Add the Text1 column to your gantt view, pick a value for your Text1 column of your task, then hide the column from the view again so it is not showing.
4. Save the file as an .xml file in MSP.
5. Import the .xml as a new project in MSP, do not show the Text1 column or contents at any time.
6. Save the project as an .mpp (e.g. TaskLookupValues**Fault**.mpp)
7. Save the project again as an .xml (e.g. TaskLookupValuesFixed.xml)
8. Save the project again as an .mpp (e.g. TaskLookupValues**Fixed**.mpp)
9. Repeat steps 7 and 8 again; for some reason you may have to save it a couple of times when using this version of the workaround.
10. Try to open both .mpp files (fixed and fault) with MPXJ and output the value of the Text1 field for the tasks

## Example outputs:
```
D:\mpxj-test>java -classpath out\production\mpxj;lib/* com.regoconsulting.mpxj.Main c:\temp\TaskLookupValuesFault.mpp
Project1: null
Task1: ???????????

D:\mpxj-test>java -classpath out\production\mpxj;lib/* com.regoconsulting.mpxj.Main c:\temp\TaskLookupValuesFixed.mpp
Project1: null
Task1: Just
```

Due to the non-ascii / non-printable chars it just shows ???? but in a non-ascii (e.g. Unicode) aware display it shows this instead, in case it "helps" with identification of where the funky data was being pulled from:
![image](https://user-images.githubusercontent.com/40403963/63405966-0eb54f80-c3ae-11e9-96c0-98ed994558f5.png)

## Expected results:
Both projects to have had the same output (e.g. "Just" for Task1).

## Workarounds:
Aside from saving-as xml/mpp a couple of times, which seems to force MSP to "read" the lookups data in a way that influences how it then stores the view/lookup/task data back into the MPP later on, it's also possible to workaround by ensuring that Text1 is added to the view inside MSP and showing for every row that has a value before saving it.

I guess in both ways, we're forcing MSP to have to acknowledge and read the data in a way that alters how it then saves it again, otherwise the place or ordering in the file for where it stores that data will be elsewhere and MPXJ picks up something else.

## Sample project files:
Here's the examples I used for the code and results above; I created this project and task manually in the MSP UI - there's nothing special about their generation.  I include the original .xml used to import as a new project, and then the subsequent saves of that file to get the bad and good .mpp's to demonstrate with.

[TaskLookupValuesDemo.zip](https://github.com/joniles/mpxj/files/3523742/TaskLookupValuesDemo.zip)

**Comment by ndarlington** [comment]
I can probably get project files like these demo ones for other versions of MSP too (such as post-June update ones, given what happened with #107 ) if needed, as I am not necessarily trusting MS to have always retained consistency in the "pre-workaround" save state.

That said, the workaround(s) appear to be universally applicable to any version of MSP, which is something.  Since we're using VSTO and MPXJ together, I can at least drive MSP a little to automate the additional save-as xml/mpp before handing off to MPXJ for the rest of our work, but it's not without its overheads especially with some of the project plan sizes we face.

**Comment by ndarlington** [comment]
My understanding of the formats involved are practically non-existent, but I do see that this looks like some sort of buffer overrun of the field due to the presence of unexpected data at that location:

![image](https://user-images.githubusercontent.com/40403963/63603397-39123300-c58f-11e9-9eab-f267411d8c0c.png)

The right panel shows the bytes that were read in to the task's record that seems to represent the data returned as output for Text1's value.

On the left it shows the data that was supposed to be in the same place.  I'm guessing that following the 01 07 sequence/marker, the next 01 00 00 00 is the integer that should be going in and pulling out the lookup value (the expected value of "Just" was the first entry in the table).

Unfortunately it doesn't seem to be as simple as finding where the 01 07 01 00 00 00 pattern has been moved to (if at all) because I couldn't find it anywhere else in the file.

**Comment by ndarlington** [comment]
Got something interesting for you.

I repeated my steps to create the bad .mpp file but this time picked the 3rd lookup value "Basic" in my list.

I then grabbed the block of bytes for the task in both bad files and compared; the only difference that appeared significant was here (highlighted):

![image](https://user-images.githubusercontent.com/40403963/63625854-88278a80-c5c6-11e9-8181-3e8cedd544bd.png)

2E - 2C = 02, the same difference as exists between the lookup positions (3 - 1  = 2).

So I changed it from 2E to 2D and opened up the .mpp in MSP, and sure enough it now showed the value "Some", the 2nd position lookup instead.

I realized after some further watching that you're monitoring for the 0x0700 mask to determine if it's a lookup table value instead of a string; but in these cases it's 0x0400 instead.

Then another observation; the 2C is the same in both the bad and the good versions of the project/task.

It was at this point I was starting to get a little excited... so I opened up the .xml to see if my suspicions were correct:

```
			<Values>
				<Value>
					<ValueID>1</ValueID>
					<FieldGUID>2C9DC596-D2C3-E911-AFF2-F48C50A1C759</FieldGUID>
					<ParentValueID>0</ParentValueID>
					<Type>21</Type>
					<IsCollapsed>0</IsCollapsed>
					<Value>Just</Value>
				</Value>
				<Value>
					<ValueID>2</ValueID>
					<FieldGUID>2D9DC596-D2C3-E911-AFF2-F48C50A1C759</FieldGUID>
					<ParentValueID>0</ParentValueID>
					<Type>21</Type>
					<IsCollapsed>0</IsCollapsed>
					<Value>Some</Value>
				</Value>
				<Value>
					<ValueID>3</ValueID>
					<FieldGUID>2E9DC596-D2C3-E911-AFF2-F48C50A1C759</FieldGUID>
					<ParentValueID>0</ParentValueID>
					<Type>21</Type>
					<IsCollapsed>0</IsCollapsed>
					<Value>Basic</Value>
				</Value>
				<Value>
					<ValueID>4</ValueID>
					<FieldGUID>AC6B6B9D-D2C3-E911-AFF2-F48C50A1C759</FieldGUID>
					<ParentValueID>0</ParentValueID>
					<Type>21</Type>
					<IsCollapsed>0</IsCollapsed>
					<Value>Lookup</Value>
				</Value>
				<Value>
					<ValueID>5</ValueID>
					<FieldGUID>AD6B6B9D-D2C3-E911-AFF2-F48C50A1C759</FieldGUID>
					<ParentValueID>0</ParentValueID>
					<Type>21</Type>
					<IsCollapsed>0</IsCollapsed>
					<Value>Values</Value>
				</Value>
			</Values>
```

Check out those GUID strings vs. the bytes in the files, good or bad.  It's the GUID right there:

![image](https://user-images.githubusercontent.com/40403963/63626219-d5582c00-c5c7-11e9-8692-a32c7ec989b3.png)

**Comment by ndarlington** [comment]
I had a crack at it.. probably naively, but it is showing some positive results in cases we've encountered

https://github.com/regoconsulting/mpxj/pull/1

**Comment by joniles** [comment]
@ndarlington thanks for the sleuthing and the suggested code changes. I did some more work locally with the sample files I have access to and have merged some changes to MPXJ which I think fix the issue. Let me know how you get on!

**Comment by ndarlington** [comment]
Thanks @joniles - I was out of town but back and keen to look into this.

I'm just being delayed a bit more than expected because I've also just hit an issue with outline code values not being readable from an MSP-saved mpp under certain conditions.  Bad news for me as that entails data loss.

However from what I've seen looking into both that issue and this fix is that they appear completely unrelated at least (and maybe more related to the bi-annual MSP channel update that just landed on 29th Sept), so once I get some more data on the other problem I'll write up a separate issue number for it.

I should have my findings wrapped up for the fix for this one (#114) in the next day or two as well and just didn't want to leave it hanging as if I wasn't also looking into it.  Thanks a lot for the changes.
